controllers/clusterpools/clusterpools-controller.go

```diff
@@ -0,0 +1,238 @@
+// Copyright Contributors to the Open Cluster Management project.
+
+package clusterpools
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const DEBUG = 1
+const INFO = 0
+const WARN = -1
+const ERROR = -2
+const FINALIZER = "clusterpools-controller.open-cluster-management.io/cleanup"
+
+const LABEL_NAMESPACE = "open-cluster-management.io/managed-by"
+const CLUSTERPOOLS = "clusterpools"
+
+// ClusterPoolsReconciler reconciles a ClusterPool, mainly for the delete
+type ClusterPoolsReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+func (r *ClusterPoolsReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+
+	ctx := context.Background()
+
+	log := r.Log.WithValues("ClusterPoolsReconciler", req.NamespacedName)
+
+	var cp hivev1.ClusterPool
+	if err := r.Get(ctx, req.NamespacedName, &cp); err != nil {
+		log.V(INFO).Info("Resource deleted")
+
+		return ctrl.Result{}, nil
+	}
+
+	target := cp.Name
+	log.V(INFO).Info("Reconcile cluster pool: " + target)
+
+	if cp.DeletionTimestamp != nil {
+		if err := deleteResources(r, &cp); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, removeFinalizer(r, &cp)
+	}
+
+	setFinalizer(r, &cp)
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClusterPoolsReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&hivev1.ClusterPool{}).WithEventFilter(predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return true
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
```

Comment:
delete should be `true` and all other should be `false` ?

Reply:
Hi @leena-jawale ,
Let me know what you think.
* I had create `true` on create, to apply the finalizer. This also causes a reconcile whenever the controller restarts, which is helpful if the controller was OFF when a clusterpools was created or delete.
* I had update because the delete is actually an `update` since there is a finalizer.  The delete flow doesn't actually happen till the resource is gone, and at that point there is no more work as the finalizer was cleared in the update before it.

For the last one, I'm also wondering if its because there is my finalizer and the hive finalizer.  When they clear there's it is an update, which triggers mine.  But when I have `update: false` and `delete: true`, it never triggers because of the finalizer.

Reply:
>I had create true on create, to apply the finalizer. This also causes a reconcile whenever the controller restarts, which is helpful if the controller was OFF when a clusterpools was created or delete.

This will need `create: true`

For 2nd point, it sounds weird. You are right when delete happens, it is adding deletiontimestamp so update will trigger.   Maybe `delete` should be above of `update` in code. So it will trigger `delete: true` when clusterpool deleted. or I am fine with whatever it is now. `create: true`,  `update: true` , `delete:false`

Reply:
I'll try flipping them, but when I had create: false, update: false and delete: true it never ran.

Reply:
Guess I should have read a little more [here](https://book.kubebuilder.io/reference/using-finalizers.html).

> The key point to note is that a finalizer causes “delete” on the object to become an “update” to set deletion timestamp. Presence of deletion timestamp on the object indicates that it is being deleted.